### PR TITLE
engineering: add grpc cs/ce/ts enablement guide

### PR DESCRIPTION
**Pull request by ggilmore**
The purpose of this guide is to inform our CS/CE/TS teams about the upcoming gRPC rollout to self-hosted instances for the upcoming `5.2.X` release. 

The biggest things I wanted to get across are:
- A general understanding of what gRPC is, what is changing, and why
- The fact that this will be enabled by default for the next release, and the circumstance that might require a customer to change their infrastructure configuration for this to work
- How to know if something is going wrong, and what to do about it.

I'll share this back in the [Slack thread](https://sourcegraph.slack.com/archives/C04DM1NAGDS/p1692998039390759) in the morning asking for people to read it over and to help identify customers we should reach out to / get on calls with before the `5.2.X` release. However, I thought it'd be good for y'all to give it a quick read over and make any necessary comments beforehand. 

Thanks!

You can see a preview of this new handbook page here: https://deploy-preview-7779--sourcegraph-handbook.netlify.app/departments/engineering/dev/tools/grpc_cs_enablement/

### CODE REVIEW

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# gRPC CS Enablement Guide
+
+# **What is gRPC?**
+
+[gRPC](https://grpc.io/) is a high-performance [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call) framework developed by Google.
+
+## What is changing?
+
+In Sourcegraph `5.2.X`, we are enabling gRPC by default (with the option to disable it in case of an issue). gRPC will be the only option in the `5.3.X` release.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Something in these two paragraphs should tell me that we use `gRPC` to communicate between internal services. That's an important bit of the context: we do NOT change external APIs, we only change the way our services communicate with each other.

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
+# gRPC CS Enablement Guide
+
+# **What is gRPC?**
+
+[gRPC](https://grpc.io/) is a high-performance [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call) framework developed by Google.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Do folks in CS know what an RPC framework is? Or should we add one sentence to explain what it is?

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+## What is changing?
+
+In Sourcegraph `5.2.X`, we are enabling gRPC by default (with the option to disable it in case of an issue). gRPC will be the only option in the `5.3.X` release.
+
+## **Why are we making this change?**
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: I'd also mention some practical benefits for our usecase: it's hard to follow references and find out exactly where an endpoint is implemented, we had multiple different representations and no standard way, when you added a new endpoint you had too many options to choose from, etc.

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+- Well-defined semantics for backwards compatibility
+- Well-typed APIs with code generation across a wide variety of programming languages
+
+See the [PR-FAQ for more background](https://docs.google.com/document/d/1lf7REi_4t03tIxSPccDG4VHtNo-6SLI1-3zDcrspi6k/edit).
+
+## What do customers need to know?
+
+gRPC **\*\*\*\***\*\*\*\***\*\*\*\***only affects traffic **_between_** our microservices (e.x. `searcher` ↔ `gitserver`). Traffic between the Sourcegraph Web UI and the rest of the application is unaffected (e.x. `[sourcegraph.example.com](http://sourcegraph.example.com)` ↔ `frontend`’s GraphQL API).
+
+**If you don’t have any security restrictions on Sourcegraph’s _internal_ traffic, your customer shouldn’t need to tweak anything.**
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: These two lines are good and very important.

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  - [repo-updater](https://github.com/sourcegraph/deploy-sourcegraph-cloud/blob/release/base/repo-updater/repo-updater.Deployment.yaml)
+- The following aspects about Sourcegraph’s networking configuration **aren’t changing**:
+  - \***\*\*\*\*\***Ports:\***\*\*\*\*\*** all Sourcegraph services will use the same ports as they were in the **5.1.X** release.
+  - **External traffic**: gRPC only affects how Sourcegraph’s microservices communicate amongst themselves - **no new external traffic is sent via gRPC**.
+  - **Service dependencies:** each Sourcegraph service will communicate with the same set of services regardless of whether gRPC is enabled.
+    - Example: `searcher` will still need to communicate with `gitserver` to fetch repository data. Whether or not gRPC is enabled doesn’t matter.
+
+## enabling / disabling GRPC
+
+gRPC will be the only means of communication used in Sourcegraph `5.3.X`. In the `5.2.x` release, you can disable the use of gRPC by:
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Can you explain whether it's enabled by default in 5.2? This reads to me like it can be disabled but doesn't explicitly say whether it is enabled by default. It'd make it explicit, maybe with a list?

```
- `5.2`: enabled by default, can be turned off via feature flag
- `5.3`: enabled by default, cannot be turned off

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+- set the `enableGRPC` experimental feature to `false` in the site configuration file:
+  ```json
+  {
+    "experimentalFeatures": {
+      "enableGRPC": false // disabled
+    }
+  }
+  ```
+- set the environment variable `SG_FEATURE_FLAG_GRPC="false"` for every service.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Natural question that comes up: what if I only set it for a few services?

Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]:
```suggestion
- OR: set the environment variable `SG_FEATURE_FLAG_GRPC="false"` for every service.
```

People usually skip reading the line before a list and jump straight into the options. Explicit `OR: ` helps make it clear that they're reading a 2nd option and not the second step.

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  - **Service dependencies:** each Sourcegraph service will communicate with the same set of services regardless of whether gRPC is enabled.
+    - Example: `searcher` will still need to communicate with `gitserver` to fetch repository data. Whether or not gRPC is enabled doesn’t matter.
+
+## enabling / disabling GRPC
+
+gRPC will be the only means of communication used in Sourcegraph `5.3.X`. In the `5.2.x` release, you can disable the use of gRPC by:
+
+### all services besides zoekt-indexserver
+
+You can either:
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: I think `zoekt-indexserver` gets a bit lost in the header here, so I'd duplicate:

```suggestion
### all services besides zoekt-indexserver

Disabling gRPC on any service that is not `zoekt-indexserver` can be done by one of these options:
```

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    "experimentalFeatures": {
+      "enableGRPC": false // disabled
+    }
+  }
+  ```
+- set the environment variable `SG_FEATURE_FLAG_GRPC="false"` for every service.
+
+### zoekt-indexserver
+
+_(zoekt-indexserver can’t read from Sourcegraph’s site configuration, so we can only use environment variables to communicate this setting)._
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: No need for `()` here.

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+Set the environment variable `GRPC_ENABLED="false"` on the zoekt-indexserver container. (See [https://github.com/sourcegraph/deploy-sourcegraph-cloud/blob/18e5f9e450878705b7a99ee7c3bcf74c3fb68514/base/indexed-search/indexed-search.StatefulSet.yaml#L105-L106](https://github.com/sourcegraph/deploy-sourcegraph-cloud/blob/18e5f9e450878705b7a99ee7c3bcf74c3fb68514/base/indexed-search/indexed-search.StatefulSet.yaml#L105-L106) for an example:
+
+```yaml
+- name: zoekt-indexserver
+          env:
+            - name: GRPC_ENABLED
+              value: "false"
+          image: docker.io/sourcegraph/search-indexer:5.2.0
+```
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Intendation is off

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+## How do you know if there might be a problem with gRPC?
+
+_See the gRPC [monitoring guide](https://handbook.sourcegraph.com/departments/engineering/dev/tools/grpc/#grpc-monitoring-guide) for more background._
+
+It’s important to understand that gRPC is merely the transport mechanism that our entire application logic is built on top of. Our existing monitoring, alerting, and reliance on user-submitted bug reports are our best tools for identifying issues with the underlying application.
+
+However, we do have two tools that can help identify issues with gRPC.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Nitpick `identify issues` is a different task than finding out `if there might be a problem` (which is the header).

**Review by mrnugget — COMMENTED** [review]
Very cool! Left some comments/suggestions. But overall this is great and helpful. 1-2 pager entry for new tech -- could be a template

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md (hunk 2 above)
Comment (ggilmore) [reply content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Added below: 

> ## Sidebar: **What is an RPC (Remote Procedure Call)?**

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md (hunk 3 above)
Comment (ggilmore) [reply content/departments/engineering/dev/tools/grpc_cs_enablement.md]: I added a few more to the list:

> - Large 3rd party plugin ecosystem
> - Better discoverability and documentation: Finding [what RPCs are available](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f28a7fab171cacefa66553d875069b8338ee7308/-/blob/internal/gitserver/v1/gitserver_grpc.pb.go?L42-59#tab=references), [where they are implemented](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f28a7fab171cacefa66553d875069b8338ee7308/-/blob/internal/gitserver/v1/gitserver_grpc.pb.go?L328:2-328:6#tab=implementations_go), and [who is calling them](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f28a7fab171cacefa66553d875069b8338ee7308/-/blob/internal/gitserver/v1/gitserver_grpc.pb.go?L46:2-46:6#tab=references) is as easy as using “find references”.

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md (hunk 5 above)
Comment (ggilmore) [reply content/departments/engineering/dev/tools/grpc_cs_enablement.md]: I added a table to make this clear:

> ## Rollout timeline
> 
> | Sourcegraph version                   | gRPC feature status                                                                   |
> | ------------------------------------- | ------------------------------------------------------------------------------------- |
> | 5.2.X (releasing October 4th, 2023)   | On by default, can be disabled via feature flag (see “enabling/disabling GRPC below”) |
> | 5.3.X (releasing December 14th, 2023) | On by default, cannot be turned off                                                   |
>

Comment (mrnugget) [reply content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Hell yes

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
+# gRPC CS Enablement Guide / Primer
+
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md:2]: How about a 1-2 line intro?

```suggestion
In 5.2 we're rolling out gRPC as a new mechanism for our internal services to communicate with each other.

No customer should notice that something has changed.

This document aims to provide background information on gRPC & how we use it and provide help in case a customer _does_ notice.
```

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+## What is gRPC?
+
+[gRPC](https://grpc.io/) is a high-performance [Remote Procedure Call (RPC)](https://en.wikipedia.org/wiki/Remote_procedure_call) framework developed by Google.
+
+### Sidebar: What is an RPC (Remote Procedure Call)?
+
+At its most basic level, an [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call) allows one program to cause a function to execute in another program (commonly on another physical machine).
+
+In simpler terms, it's a way for different parts of a system, which could be located on different machines, to talk to each other and perform tasks.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]:
```suggestion
It's a way for different parts of a system, which could be located on different machines, to talk to each other and perform tasks.
```

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+Imagine being able to “press a button in one” application to get information or perform an action in another application located somewhere else; that's the essence of an [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call).
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]:
```suggestion
Imagine being able to “press a button” in one application to perform an action in another application located somewhere else; that's the essence of an [RPC](https://en.wikipedia.org/wiki/Remote_procedure_call).
```

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+/// ---------------
+
+// On Client Machine
+result = remoteCall('add', 5, 3) // Makes an RPC to the 'add' function on the server
+print(result) // Output will be 8
+```
+
+## What is changing?
+
+Starting in Sourcegraph `5.2.X` , we are switching to use gRPC for Sourcegraph’s **_internal communication (service to service only)_** by default.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md:38]: How about making it concrete here?

```suggestion
Starting in Sourcegraph `5.2.X` , we are switching to use gRPC for Sourcegraph’s **_internal communication (service to service only)_** by default.

In layman's terms: instead of JSON being sent via HTTP between `repo-updater` and `gitserver`, both services now use gRPC to talk to each other.
```

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+### Rollout timeline
+
+| Sourcegraph Version                   | gRPC Feature Status                                                                   |
+| ------------------------------------- | ------------------------------------------------------------------------------------- |
+| 5.2.X (releasing October 4th, 2023)   | On by default, can be disabled via feature flag (see “enabling/disabling GRPC below”) |
+| 5.3.X (releasing December 14th, 2023) | On by default, cannot be turned off                                                   |
+
+## Why are we making this change?
+
+As opposed to [REST](https://en.wikipedia.org/wiki/Representational_state_transfer) (the pattern we currently use for [RPCs](https://en.wikipedia.org/wiki/Remote_procedure_call)), gRPC offers the following benefits and more:
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Maybe worth mentioning in `()` that we don't even use REST consistently... Or that we don't have any consistent APIs across all services ;)

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+See the [PR-FAQ for more background](https://docs.google.com/document/d/1lf7REi_4t03tIxSPccDG4VHtNo-6SLI1-3zDcrspi6k/edit).
+
+## What do customers need to know?
+
+### Most important takeaway
+
+The most important takeaway is this:
+
+> **If you don’t have any security restrictions on Sourcegraph’s _internal_ traffic, your customer shouldn’t need to tweak anything.**
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Sorry, but pet peeve of mine is using quotes to highlight text :)

```suggestion
**If you don’t have any security restrictions on Sourcegraph’s _internal_ traffic, your customer shouldn’t need to tweak anything.**
```

File: content/departments/engineering/dev/tools/grpc_cs_enablement.md [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+If you see either:
+
+- relevant entries in the “GRPC “internal errors” metrics**”** grafana dashboards
+- **OR**: application logs that are prefixed with `grpc.internal.error.reporter`
+
+That indicates a surefire issue with our gRPC implementation and (should be reported to us).
+
+## Who do you contact if there is a problem?
+
+Contact us in the #job-fair-grpc Slack channel.
```
Comment (mrnugget) [inline content/departments/engineering/dev/tools/grpc_cs_enablement.md]: Should we maybe add `#team-apis` too? Just to make sure that this here outlives the channel in case we decide after Job Fair ends to rename/archive it.

**Review by mrnugget — APPROVED** [review]
Very nice!
